File: abstractions/php/src/Serialization/SerializationWriter.php
<?php
namespace Microsoft\Kiota\Abstractions\Serialization;

use DateInterval;
use DateTime;
use Microsoft\Kiota\Abstractions\Enum;
use Microsoft\Kiota\Abstractions\Types\Byte;
use Microsoft\Kiota\Abstractions\Types\Date;
use Microsoft\Kiota\Abstractions\Types\Time;
use Psr\Http\Message\StreamInterface;

/** Defines an interface for serialization of objects to a stream. */
interface SerializationWriter {
    /**
     * Writes the specified string value to the stream with an optional given key.
     * @param string|null $key the key to write the value with.
     * @param string|null $value the value to write to the stream.
     */
    public function writeStringValue(?string $key, ?string $value): void;

    /**
     * Writes the specified Boolean value to the stream with an optional given key.
     * @param string|null $key the key to write the value with.
     * @param bool|null $value the value to write to the stream.
     */
    public function writeBooleanValue(?string $key, ?bool $value): void;

    /**
     * Writes the specified Float value to the stream with an optional given key.
     * @param string|null $key the key to write the value with.
     * @param float|null $value the value to write to the stream.
     */
    public function writeFloatValue(?string $key, ?float $value): void;

    /**
     * Writes the specified Integer value to the stream with an optional given key.
     * @param string|null $key the key to write the value with.
     * @param int|null $value the value to write to the stream.
     */
    public function writeIntegerValue(?string $key, ?int $value): void;

    /**
     * Writes the specified OffsetDateTime value to the stream with an optional given key.
     * @param string|null $key the key to write the value with.
     * @param DateTime|null $value the value to write to the stream.
     */
    public function writeDateTimeValue(?string $key, ?DateTime $value): void;

    /**
     * Writes the specified collection of object values to the stream with an optional given key.
     * @param string|null $key the key to write the value with.
     * @param array<Parsable>|null $values
     */
    public function writeCollectionOfObjectValues(?string $key, ?array $values): void;

    /**
     * Writes the specified model object value to the stream with an optional given key.
     * @param string|null $key the key to write the value with.
     * @param Parsable|null $value the value to write to the stream.
     */
    public function writeObjectValue(?string $key, ?Parsable $value): void;

    /**
     * Gets the value of the serialized content.
     * @return StreamInterface the value of the serialized content.
     */
    public function getSerializedContent(): StreamInterface;

    /**
     * Writes the specified enum set value to the stream with an optional given key.
     * @param string|null $key the key to write the value with.
     * @param array<Enum>|null $values the value to write to the stream.
     */
    public function writeEnumSetValue(?string $key, ?array $values): void;

    /**
     * Writes the specified enum value to the stream with an optional given key.
     * @param string|null $key the key to write the value with.
     * @param Enum|null $value the value to write to the stream.
     */
    public function writeEnumValue(?string $key, ?Enum $value): void;

    /**
     * Writes a null value for the specified key.
     * @param string|null $key the key to write the value with.
     */
    public function writeNullValue(?string $key): void;

    /**
     * Writes the specified additional data values to the stream with an optional given key.
     * @param array<string,mixed> $value the values to write to the stream.
     */
    public function writeAdditionalData(?array $value): void;

    /**
     * Write the Date-only Segment of DateTime.
     * @param string|null $key
     * @param Date|null $value
     */
    public function writeDateValue(?string $key, ?Date $value): void;

    /**
     * Write a TimeOnly value without the
     * @param string|null $key
     * @param Time|null $value
     * @return void
     */
    public function writeTimeValue(?string $key, ?Time $value): void;

    /**
     * Write DateInterval Value.
     * @param string|null $key
     * @param DateInterval|null $value
     * @return void
     */
    public function writeDateIntervalValue(?string $key, ?DateInterval $value): void;
    /**
     * Write a byte value.
     * @param string|null $key
     * @param Byte|null $value
     * @return void
     */
    public function writeByteValue(?string $key, ?Byte $value): void;

    /**
     * @param string|null $key
     * @param array<mixed>|null $value
     * @return void
     */
    public function writeCollectionOfPrimitiveValues(?string $key, ?array $value): void;
    
    /**
    * Try to write a value with any type.
    * @param string|null $key The key for the value.
    * @param mixed|null $value.
    **/
    public function writeAnyValue(?string $key, $value): void;

    /**
     * Write a binary stream.
     * @param string|null $key
     * @param StreamInterface|null $value
     * @return void
     */
    public function writeBinaryContent(?string $key, ?StreamInterface $value): void;
    
    /**
     * Sets the callback called before the objects gets serialized.
     * @param callable|null $value the callback called before the objects gets serialized.
     */
    public function setOnBeforeObjectSerialization(?callable $value): void;

    /**
     * Gets the callback called before the object gets serialized.
     * @return callable|null the callback called before the object gets serialized.
     */
    public function getOnBeforeObjectSerialization(): ?callable;

    /**
     * Sets the callback called after the objects gets serialized.
     * @param callable|null $value the callback called after the objects gets serialized.
     */
    public function setOnAfterObjectSerialization(?callable $value): void;

    /**
     * Gets the callback called after the object gets serialized.
     * @return callable|null the callback called after the object gets serialized.
     */
    public function getOnAfterObjectSerialization(): ?callable;

    /**
     * Sets the callback called right after the serialization process starts.
     * @param callable|null $value the callback called right after the serialization process starts.
     */
    public function setOnStartObjectSerialization(?callable $value): void;

    /**
     * Gets the callback called right after the serialization process starts.
     * @return callable|null the callback called right after the serialization process starts.
     */
    public function getOnStartObjectSerialization(): ?callable;
}

File: abstractions/go/serialization/serialization_writer.go
package serialization

import (
	i "io"
	"time"

	"github.com/google/uuid"
)

// Defines an interface for serialization of objects to a byte array.
type SerializationWriter interface {
	i.Closer
	// Writes a String value to the byte array.
	// Parameters:
	// - key - the key of the value to write (optional).
	// - value - the String value to write.
	// Returns:
	// - An error if any.
	WriteStringValue(key string, value *string) error
	// Writes a Bool value to the byte array.
	// Parameters:
	// - key - the key of the value to write (optional).
	// - value - the Bool value to write.
	// Returns:
	// - An error if any.
	WriteBoolValue(key string, value *bool) error
	// Writes a Int32 value to the byte array.
	// Parameters:
	// - key - the key of the value to write (optional).
	// - value - the Int32 value to write.
	// Returns:
	// - An error if any.
	WriteInt32Value(key string, value *int32) error
	// Writes a Int64 value to the byte array.
	// Parameters:
	// - key - the key of the value to write (optional).
	// - value - the Int64 value to write.
	// Returns:
	// - An error if any.
	WriteInt64Value(key string, value *int64) error
	// Writes a Float32 value to the byte array.
	// Parameters:
	// - key - the key of the value to write (optional).
	// - value - the Float32 value to write.
	// Returns:
	// - An error if any.
	WriteFloat32Value(key string, value *float32) error
	// Writes a Float64 value to the byte array.
	// Parameters:
	// - key - the key of the value to write (optional).
	// - value - the Float64 value to write.
	// Returns:
	// - An error if any.
	WriteFloat64Value(key string, value *float64) error
	// Writes a ByteArray value to the byte array.
	// Parameters:
	// - key - the key of the value to write (optional).
	// - value - the ByteArray value to write.
	// Returns:
	// - An error if any.
	WriteByteArrayValue(key string, value []byte) error
	// Writes a Time value to the byte array.
	// Parameters:
	// - key - the key of the value to write (optional).
	// - value - the Time value to write.
	// Returns:
	// - An error if any.
	WriteTimeValue(key string, value *time.Time) error
	// Writes a UUID value to the byte array.
	// Parameters:
	// - key - the key of the value to write (optional).
	// - value - the UUID value to write.
	// Returns:
	// - An error if any.
	WriteUUIDValue(key string, value *uuid.UUID) error
	// Writes a Parsable value to the byte array.
	// Parameters:
	// - key - the key of the value to write (optional).
	// - value - the Parsable value to write.
	// Returns:
	// - An error if any.
	WriteObjectValue(key string, item Parsable) error
	// Writes a collection of Parsable values to the byte array.
	// Parameters:
	// - key - the key of the value to write (optional).
	// - collection - the collection of Parsable value to write.
	// Returns:
	// - An error if any.
	WriteCollectionOfObjectValues(key string, collection []Parsable) error
	// Writes a collection of String values to the byte array.
	// Parameters:
	// - key - the key to write (optional).
	// - collection - the collection to write.
	// Returns:
	// - An error if any.
	WriteCollectionOfStringValues(key string, collection []string) error
	// Writes a collection of Bool values to the byte array.
	// Parameters:
	// - key - the key to write (optional).
	// - collection - the collection to write.
	// Returns:
	// - An error if any.
	WriteCollectionOfBoolValues(key string, collection []bool) error
	// Writes a collection of Int32 values to the byte array.
	// Parameters:
	// - key - the key to write (optional).
	// - collection - the collection to write.
	// Returns:
	// - An error if any.
	WriteCollectionOfInt32Values(key string, collection []int32) error
	// Writes a collection of Int64 values to the byte array.
	// Parameters:
	// - key - the key to write (optional).
	// - collection - the collection to write.
	// Returns:
	// - An error if any.
	WriteCollectionOfInt64Values(key string, collection []int64) error
	// Writes a collection of Float32 values to the byte array.
	// Parameters:
	// - key - the key to write (optional).
	// - collection - the collection to write.
	// Returns:
	// - An error if any.
	WriteCollectionOfFloat32Values(key string, collection []float32) error
	// Writes a collection of Float64 values to the byte array.
	// Parameters:
	// - key - the key to write (optional).
	// - collection - the collection to write.
	// Returns:
	// - An error if any.
	WriteCollectionOfFloat64Values(key string, collection []float64) error
	// Writes a collection of Time values to the byte array.
	// Parameters:
	// - key - the key to write (optional).
	// - collection - the collection to write.
	// Returns:
	// - An error if any.
	WriteCollectionOfTimeValues(key string, collection []time.Time) error
	// Writes a collection of UUID values to the byte array.
	// Parameters:
	// - key - the key to write (optional).
	// - collection - the collection to write.
	// Returns:
	// - An error if any.
	WriteCollectionOfUUIDValues(key string, collection []uuid.UUID) error
	// Gets the resulting byte array from the serialization writer.
	// Returns:
	// - The resulting byte array.
	// - An error if any.
	GetSerializedContent() ([]byte, error)
	// Writes additional data to the byte array.
	// Parameters:
	// - value - the data to write.
	// Returns:
	// - An error if any.
	WriteAdditionalData(value map[string]interface{}) error
}
